### PR TITLE
http: make max-filesize check not count ignored bodies

### LINKED ISSUE #14899: HTTP --max-filesize counts ignored body data
From https://github.com/curl/curl/issues/14440#issuecomment-2346734573 user @MasterInQuestion reports that --max-filesize is applied to the ignored content of a redirect response. For example `curl -v -L --max-filesize 1 http://localhost:8080` and the server's redirect reply has 5 content bytes:

~~~
while true; \
do perl -e 'print ("HTTP/1.1 301\r\nContent-Length: 5\r\nLocation: http://localhost:8080/abc\r\n\r\napple")' | nc -4l localhost 8080; \
done
~~~

curl will fail because 5 bytes of content (even though it's ignored) is more than 1:

~~~
* STATE: DO => DID handle 0x76bd88; line 2164
* STATE: DID => PERFORMING handle 0x76bd88; line 2282
< HTTP/1.1 301
< Content-Length: 5
< Location: http://localhost:8080/abc
* Maximum file size exceeded
~~~

The failure is caused by a check in `Curl_http_size` which AFAICT compares even for body bytes that are not "real" so to speak:

https://github.com/curl/curl/blob/a3bd1dda12ec79cd63e0d81df4ec4b1fbbbcfa1b/lib/http.c#L3273-L3294

There is a separate check that is done in `cw_download_write` which deals with real body bytes, however by that point the download is already taking place.

https://github.com/curl/curl/blob/a3bd1dda12ec79cd63e0d81df4ec4b1fbbbcfa1b/lib/sendf.c#L326-L344

In other words, the former check causes curl to error immediately after the headers are received and the content length is known. The user may want this behavior because if they, say, limit to 1MB then maybe they don't want to download the first 1MB of a 100MB file. OTOH it seems that cw_download_write works in exactly that way, it will allow downloading to the maximum file size.

So if the former check is removed, which is my initial inclination, there will be a change in behavior.

### PR DESCRIPTION
**Pull request by bagder**
Add test 477 to verify

Reported-by: Jay Satiro
Fixes #14899

### CODE REVIEW

**Comment by dfandrich** [comment]
Analysis of PR #14900 at [90397d5b](https://github.com/curl/curl/commit/90397d5b2efa50b92170e8a187759ae462153777):

[Test 477 failed](https://github.com/curl/curl/actions/runs/10852283724/job/30117991825#check-step-41), which has NOT been flaky recently, so **there could be a real issue in the PR**.

[Test 2055 failed](https://github.com/curl/curl/actions/runs/10852283723/job/30117989141#check-step-13), which has NOT been flaky recently, so **there could be a real issue in the PR**.

###### Generated by [Testclutch](https://testclutch.curl.se/)

File: lib/http.c [hunk 1]
```diff
@@ -3283,10 +3283,13 @@ CURLcode Curl_http_size(struct Curl_easy *data)
   }
   else if(k->size != -1) {
     if(data->set.max_filesize &&
-       k->size > data->set.max_filesize) {
-      failf(data, "Maximum file size exceeded");
+       !k->ignorebody &&
+       (k->size > data->set.max_filesize)) {
+      failf(data, "Maximum **file size exceeded");
```
Comment (jay) [inline lib/http.c:3288]:
```suggestion
      failf(data, "Maximum **file size exceeded");
```

the old text is better it doesn't need ***

**Comment by jay** [comment]
test 477 failure of hyper and max filesize looks legit

~~~
2024-09-13T16:00:51.9548086Z  0000: HTTP/1.1 301 OK
2024-09-13T16:00:51.9548855Z  16:00:51.225896 [0-0] == Info: [WRITE] cw_out, wrote 17 header bytes -> 17
2024-09-13T16:00:51.9550054Z  16:00:51.225899 [0-0] == Info: [WRITE] download_write header(type=c, blen=17) -> 0
2024-09-13T16:00:51.9550836Z  16:00:51.225902 [0-0] == Info: [WRITE] client_write(type=c, len=17) -> 0
2024-09-13T16:00:51.9551517Z  16:00:51.225912 [0-0] <= Recv header, 37 bytes (0x25)
2024-09-13T16:00:51.9552208Z  0000: Date: Tue, 09 Nov 2010 14:49:00 GMT
2024-09-13T16:00:51.9553194Z  16:00:51.225923 [0-0] == Info: [WRITE] header_collect pushed(type=1, len=37) -> 0
2024-09-13T16:00:51.9554766Z  16:00:51.225927 [0-0] == Info: [WRITE] cw_out, wrote 37 header bytes -> 37
2024-09-13T16:00:51.9556190Z  16:00:51.225929 [0-0] == Info: [WRITE] download_write header(type=4, blen=37) -> 0
2024-09-13T16:00:51.9557264Z  16:00:51.225931 [0-0] == Info: [WRITE] client_write(type=4, len=37) -> 0
2024-09-13T16:00:51.9558274Z  16:00:51.225935 [0-0] <= Recv header, 26 bytes (0x1a)
2024-09-13T16:00:51.9558997Z  0000: Server: test-server/fake
2024-09-13T16:00:51.9559881Z  16:00:51.225943 [0-0] == Info: [WRITE] header_collect pushed(type=1, len=26) -> 0
2024-09-13T16:00:51.9561017Z  16:00:51.225945 [0-0] == Info: [WRITE] cw_out, wrote 26 header bytes -> 26
2024-09-13T16:00:51.9562185Z  16:00:51.225948 [0-0] == Info: [WRITE] download_write header(type=4, blen=26) -> 0
2024-09-13T16:00:51.9563335Z  16:00:51.225950 [0-0] == Info: [WRITE] client_write(type=4, len=26) -> 0
2024-09-13T16:00:51.9564282Z  16:00:51.225956 [0-0] <= Recv header, 19 bytes (0x13)
2024-09-13T16:00:51.9564940Z  0000: Location: 4770002
2024-09-13T16:00:51.9565499Z  16:00:51.225963 [0-0] == Info: [WRITE] header_collect pushed(type=1, len=19) -> 0
2024-09-13T16:00:51.9566156Z  16:00:51.225966 [0-0] == Info: [WRITE] cw_out, wrote 19 header bytes -> 19
2024-09-13T16:00:51.9566824Z  16:00:51.225968 [0-0] == Info: [WRITE] download_write header(type=4, blen=19) -> 0
2024-09-13T16:00:51.9567644Z  16:00:51.225970 [0-0] == Info: [WRITE] client_write(type=4, len=19) -> 0
2024-09-13T16:00:51.9568179Z  16:00:51.225973 [0-0] <= Recv header, 22 bytes (0x16)
2024-09-13T16:00:51.9568570Z  0000: Accept-Ranges: bytes
2024-09-13T16:00:51.9569426Z  16:00:51.225980 [0-0] == Info: [WRITE] header_collect pushed(type=1, len=22) -> 0
2024-09-13T16:00:51.9570569Z  16:00:51.225983 [0-0] == Info: [WRITE] cw_out, wrote 22 header bytes -> 22
2024-09-13T16:00:51.9571708Z  16:00:51.225985 [0-0] == Info: [WRITE] download_write header(type=4, blen=22) -> 0
2024-09-13T16:00:51.9572370Z  16:00:51.225986 [0-0] == Info: [WRITE] client_write(type=4, len=22) -> 0
2024-09-13T16:00:51.9572900Z  16:00:51.225990 [0-0] <= Recv header, 20 bytes (0x14)
2024-09-13T16:00:51.9573285Z  0000: Content-Length: 26
2024-09-13T16:00:51.9573760Z  16:00:51.225998 [0-0] == Info: [WRITE] header_collect pushed(type=1, len=20) -> 0
2024-09-13T16:00:51.9574536Z  16:00:51.226000 [0-0] == Info: [WRITE] cw_out, wrote 20 header bytes -> 20
2024-09-13T16:00:51.9575182Z  16:00:51.226002 [0-0] == Info: [WRITE] download_write header(type=4, blen=20) -> 0
2024-09-13T16:00:51.9575821Z  16:00:51.226004 [0-0] == Info: [WRITE] client_write(type=4, len=20) -> 0
2024-09-13T16:00:51.9576488Z  16:00:51.226007 [0-0] <= Recv header, 20 bytes (0x14)
2024-09-13T16:00:51.9576886Z  0000: Funny-head: yesyes
2024-09-13T16:00:51.9577350Z  16:00:51.226021 [0-0] == Info: [WRITE] header_collect pushed(type=1, len=20) -> 0
2024-09-13T16:00:51.9577992Z  16:00:51.226023 [0-0] == Info: [WRITE] cw_out, wrote 20 header bytes -> 20
2024-09-13T16:00:51.9578782Z  16:00:51.226025 [0-0] == Info: [WRITE] download_write header(type=4, blen=20) -> 0
2024-09-13T16:00:51.9579415Z  16:00:51.226027 [0-0] == Info: [WRITE] client_write(type=4, len=20) -> 0
2024-09-13T16:00:51.9579972Z  16:00:51.226030 [0-0] == Info: Maximum **file size exceeded
~~~

File: lib/http.c (hunk 1 above)
Comment (bagder) [reply lib/http.c]: that was a mistake debug leftover

**Comment by bagder** [comment]
I'm disable this test for hyper (mostly since hyper is already scheduled for removal). In the hyper situation `ignorebody` is not set at the time of the check.

**Review by jay — APPROVED** [review]
I would change reported line to MasterInQuestion on github since he is the one who originally reported it

**Pull request closed without merging**